File: SMSlinregpull.m
% ---------------------------------------------------------------------
% Book:         SMS
% ---------------------------------------------------------------------
% Quantlet:     SMSlinregpull
% ---------------------------------------------------------------------
% Description:  SMSlinregpull computes a linear regression of sales (X1) 
%               on price (X2) from the pullovers data set 
%               (pullover.dat).
% ---------------------------------------------------------------------
% Usage:        -
% ---------------------------------------------------------------------
% Inputs:       None
% ---------------------------------------------------------------------
% Output:       Linear regression of sales (X1) on price (X2) from the
%               pullovers data set (pullover.dat).
% ---------------------------------------------------------------------
% Example:      
% ---------------------------------------------------------------------
% Author:       Norman Duckwitz 20140220
% ---------------------------------------------------------------------

clear all
close all
clc

load('pullover.dat');           % load data
x     = pullover;
[m,n] = size(x);
sales = x(:,1);
price = x(:,2);

 % Construct X matrix
ones  = ones(m,1);
X     = [ones x(:,2)];


 % Regresssion Sales ~ Price whole set
whichstats = {'standres','r','beta'};
stats      = regstats(sales,price,'linear',whichstats);
stdres     = stats.standres;
beta       = stats.beta;
r          = stats.r;

outlier    = (abs(stdres)>2)

 % Regresssion Sales ~ Price without outliers
whichstats = {'standres','r','beta'};
stats      = regstats(sales(outlier~= 1),price(outlier~= 1),'linear',whichstats);
stdreso    = stats.standres;
betao      = stats.beta;
ro         = stats.r;

xmin       = min(price);
xmax       = max(price)+2;
ymin       = min(sales)-2;
ymax       = max(sales)+2;

 % Plot

subplot(1,2,1)
scatter(price(outlier~= 1),sales(outlier~= 1),'k');
hold on
scatter(price(outlier== 1),sales(outlier== 1),'^','filled','red');
line(price,X*beta,'Color','b','LineStyle','-','LineWidth',0.5);
line(price(outlier~= 1),X(outlier~= 1,:)*betao,'Color','black','LineStyle','--','LineWidth',0.5);
axis([xmin xmax ymin ymax]);
xlabel('Price');
ylabel('Sales');
title('Regression with Outliers');
box on
hold off

subplot(1,2,2);
boxplot(r,'Symbol','o','labels',{''});
title('Boxplot of Residuals');

 print -painters -dpdf -r600 SMSlinregpull.pdf
 print -painters -dpng -r600 SMSlinregpull.png

File: SMSlinregpull.r
# ---------------------------------------------------------------------
# Book:         SMS
# ---------------------------------------------------------------------
# Quantlet:     SMSlinregpull
# ---------------------------------------------------------------------
# Description:  SMSlinregpull computes a linear regression of sales (X1) 
#               on price (X2) from the pullovers data set 
#               (pullover.rda). It checks for outliers using a rule on
#               standardized residuals and computes the same model for
#               the subset of the pullover data without outliers. Both
#               regression lines are plotted in the same scatterplot,
#               the outliers are marked as red triangles.
# ---------------------------------------------------------------------
# See also:     SMSanovapull, SMSdete2pull, SMSdeterpull, SMSlinregpull,
#               SMSscabank45, SMScovbank
# ---------------------------------------------------------------------
# Keywords:     linear, linear model, linear regression, least-squares, R-squared
#               regression, F test, F-statistic, F-test, test, summary,
#               statistics, boxplot, five number summary, plot,
#               data visualization, graphical representation,
#               visualization, scatterplot, 
# ---------------------------------------------------------------------
# Usage:        -
# ---------------------------------------------------------------------
# Inputs:       None
# ---------------------------------------------------------------------
# Output:       Linear regression of sales (X1) on price (X2) from the
#               pullovers data set (pullover.rda).
# ---------------------------------------------------------------------
# Example:      -
# ---------------------------------------------------------------------
# Author:       Zdenek Hlavka
# ---------------------------------------------------------------------

# Clear all variables and close windows
 rm(list=ls(all=TRUE))
 graphics.off()

# setwd("C:/...") # please change your working directory

# install and load packages
libraries = c("MASS")
lapply(libraries, function(x) if (!(x %in% installed.packages())) {
  install.packages(x)
})
lapply(libraries, library, quietly = TRUE, character.only = TRUE)

load("pullover.rda")                       # load data

lm.pull = lm( Sales~Price,data=pullover)   # linear model, y=Sales, X=[Price] including intercept 
opar    = par(mfrow=c(1,2))                # define screen with 1 row and 2 columns
outlier = (abs(stdres(lm.pull))>2)         # check for outliers using a rule on standardized residuals from the regression model

# plot all data points and the regression line
plot(Sales~Price,   data=pullover, type="n",main="Regression with outliers")
points(Sales~Price, data=pullover, subset=!outlier)
points(Sales~Price, data=pullover, subset=outlier,col="red",pch=17)
abline(coef(lm.pull),lty=3)

# re-do the analysis without outliers
lm.pull.2 = lm(Sales~Price, data=pullover, subset=!outlier) # linear model, y=Sales, X=[Price], on subset without outliers
abline(coef(lm.pull.2),col="blue")

# boxplot of the original residuals
boxplot(lm.pull$residuals,main="Boxplot of residuals")
par(opar)
